skimage/feature/_hog.py

```diff
@@ -97,19 +124,34 @@ def hog(image, orientations=9, pixels_per_cell=(8, 8),
     e.g. bar like structures in bicycles and limbs in humans.
     """
 
+
     if image.dtype.kind == 'u':
         # convert uint image to float
         # to avoid problems with subtracting unsigned numbers in np.diff()
         image = image.astype('float')
 
-    gx = np.empty(image.shape, dtype=np.double)
-    gx[:, 0] = 0
-    gx[:, -1] = 0
-    gx[:, 1:-1] = image[:, 2:] - image[:, :-2]
-    gy = np.empty(image.shape, dtype=np.double)
-    gy[0, :] = 0
-    gy[-1, :] = 0
-    gy[1:-1, :] = image[2:, :] - image[:-2, :]
+    if len(image.shape) == 3:
+
+        channel_gradients = []
+        gradient = np.zeros(image.shape)
+
+        for dimension in range(image.shape[2]):
+            gx, gy = compute_image_gradients(image[:, :, dimension])
+            gradient[:, :, dimension] = np.hypot(gx, gy)
+            channel_gradients.append((gx, gy))
+
+        gx = np.empty(image.shape[0:2], dtype=np.double)
+        gy = np.empty(image.shape[0:2], dtype=np.double)
+
+        max_index_array = gradient.argmax(axis = 2)
+
```

Comment:
If you want to iterate over arrays, you should use Cython, with that said this can be done in numpy itself. I believe you are trying to index one array with the maximum index along 
the 3rd axis found elsewhere

```python
import numpy as np

a = np.random.random((10, 10, 3))*10
a = a.astype(np.int)

index = np.random.random((10, 10))*3
index = index.astype(np.int)
rr, cc = np.meshgrid(np.arange(a.shape[0]), np.arange(a.shape[1]))
b = a[rr, cc, index]

print(b)
```